framework/src/org/checkerframework/framework/flow/CFAbstractValue.java

```diff
@@ -4,550 +4,481 @@
+            this.bSet = bSet;
+            this.aAtv = getEffectTypeVar(aTypeMirror);
+            this.bAtv = getEffectTypeVar(bTypeMirror);
         }
 
-        return source;
-    }
-
-    /**
-     * For the given intersection type in the hierarchy of top, find the direct supertype that has the
-     * most specific annotation.
-     */
-    private static AnnotatedTypeMirror mostSpecificSupertype(
-            final QualifierHierarchy qualifierHierarchy,
-            final AnnotatedIntersectionType isect,
-            final AnnotationMirror top) {
-        AnnotatedTypeMirror result = null;
-        AnnotationMirror anno = null;
-        for (final AnnotatedTypeMirror supertype : isect.directSuperTypes()) {
-            final AnnotationMirror superAnno = supertype.getAnnotationInHierarchy(top);
-            if (superAnno != null
-                    && (anno == null || qualifierHierarchy.isSubtype(superAnno, anno))) {
-                anno = superAnno;
-                result = supertype;
+        void visit() {
+            Set<? extends AnnotationMirror> tops = hierarchy.getTopAnnotations();
+            for (AnnotationMirror top : tops) {
+                AnnotationMirror a = hierarchy.findAnnotationInHierarchy(aSet, top);
+                AnnotationMirror b = hierarchy.findAnnotationInHierarchy(bSet, top);
+                if (a != null && b != null) {
+                    visitAnnotationExistInBothSets(a, b, top);
+                } else if (a != null) {
+                    visitAnnotationExistInOneSet(a, bAtv, top);
+                } else if (b != null) {
+                    visitAnnotationExistInOneSet(b, aAtv, top);
+                } else {
+                    visitNeitherAnnotationExistsInBothSets(aAtv, bAtv, top);
+                }
             }
         }
 
-        return result;
-    }
+        protected abstract void visitAnnotationExistInBothSets(
+                AnnotationMirror a, AnnotationMirror b, AnnotationMirror top);
 
-    @Override
-    public boolean equals(Object obj) {
-        if (obj == null || !(obj instanceof CFAbstractValue)) {
-            return false;
-        }
-        @SuppressWarnings("rawtypes")
-        boolean result = getType().equals(((CFAbstractValue) obj).getType());
-        return result;
-    }
+        protected abstract void visitNeitherAnnotationExistsInBothSets(
+                AnnotatedTypeVariable aAtv, AnnotatedTypeVariable bAtv, AnnotationMirror top);
 
-    @Pure
-    @Override
-    public int hashCode() {
-        return HashCodeUtils.hash(type);
+        protected abstract void visitAnnotationExistInOneSet(
+                AnnotationMirror anno, AnnotatedTypeVariable atv, AnnotationMirror top);
     }
 
-    /**
-     * @return the string representation as a comma-separated list
-     */
-    @SideEffectFree
-    @Override
-    public String toString() {
-        return getType().toString();
+    private AnnotatedTypeVariable getEffectTypeVar(TypeMirror typeMirror) {
```

Comment:
A little documentation would be nice for many of these methods. What is an `EffectTypeVar`?